privacyidea/lib/tokens/certificatetoken.py

```diff
@@ -33,22 +35,176 @@
+jibBrZvzkcoJbwIDAQABoykwJzARBgorBgEEAYLECgMDBAMFAQIwEgYDVR0TAQH/
+BAgwBgEB/wIBADANBgkqhkiG9w0BAQsFAAOCAQEABVe3v1pBdPlf7C7SuHgm5e9P
+6r9aZMnPBn/KjAr8Gkcc1qztyFtUcgCfuFmrcyWy1gKjWYMxae7BXz3yKxsiyrb8
++fshMp4I8whUbckmEEIIHTy18FqxmNRo3JHx05FUeqA0i/Zl6xOfOhy/Q8XR0DMj
+xiWgTOTpqlmA2AIesBBfuOENDDAccukRMLTXDY97L2luDFTJ99NaHKjjhCOLHmTS
+BuCSS5h/F6roSi7rjh2bEOaErLG+G1hVWAzfBNXgrsUuirWRk0WKhJKeUrfNPILh
+CzYH/zLEaIsLjRg+tnmKJu2E4IdodScri7oGVhVyhUW5DrcX+/8CPqnoBpd7zQ==
+-----END CERTIFICATE-----
+"""
+
+DEFAULT_CA_PATH = ["/etc/privacyidea/trusted_attestation_ca"]
+
+
+class ACTION(BASE_ACTION):
+    __doc__ = """This is the list of special certificate actions."""
+    TRUSTED_CA_PATH = "trusted_Attestation_CA_path"
+    REQUIRE_ATTESTATION = "require_attestation"
+
+
+class REQUIRE_ACTIONS(object):
+    IGNORE = "ignore"
+    VERIFY = "verify"
+    REQUIRE_AND_VERIFY = "require_and_verify"
+
+
+def _verify_cert(parent, child):
+    ca = load_pem_x509_certificate(to_byte_string(parent), default_backend())
+    cert = load_pem_x509_certificate(to_byte_string(child), default_backend())
+    ca.public_key().verify(
+        cert.signature,
+        cert.tbs_certificate_bytes,
+        padding.PKCS1v15(),
+        cert.signature_hash_algorithm
+    )
+
+
+def verify_certificate_path(certificate, trusted_ca_paths):
+    """
+    Verify a certificate against the list of directories each containing files with
+    a certificate chain.
+
+    :param certificate: The PEM certificate to verify
+    :param trusted_ca_paths: A list of directories
+    :return: True or False
+    """
+    from os import listdir
+    from os.path import isfile, join, isdir
+    verified = False
+    for capath in trusted_ca_paths:
+        if isdir(capath):
+            chainfiles = [join(capath, f) for f in listdir(capath) if isfile(join(capath, f))]
+            for chainfile in chainfiles:
+                chain = parse_chainfile(chainfile)
+                try:
+                    verify_certificate(to_byte_string(certificate), chain)
+                    verified = True
+                except Exception as exx:
+                    log.debug(u"Can not verify attestation certificate against chain {0!s}.".format(chain))
+        else:
+            log.warning("The configured attestation CA directory does not exist.")
+    return verified
+
+
+def parse_chainfile(chainfile):
+    """
+    Parse a text file, that contains a list of CA files.
+    The topmost being the trusted Root CA followed by intermediate
+
+    :param chainfile: The filename to parse
+    :return: A list of PEM certificates
+    """
+    cacerts = []
+    cacert = ""
+    with open(chainfile) as f:
+        lines = f.readlines()
+    for line in lines:
+        if line.startswith("-----BEGIN CERTIFICATE-----"):
+            cacert = line
+        elif line.startswith("-----END CERTIFICATE-----"):
+            cacert += line
+            # End of certificate
+            cacerts.append(cacert)
```

Comment:
we could load the certificate here already so we don't have to load them several times in `_verify_cert()`.
We would also have a check if the certificate in the chain are valid.

Reply:
You are right, if there is a chain of *root*, *intermediate* and *cert*, then we would call the ``load_pem_x509_certificate`` twice for the intermediate, one time for being the signee and the second time for being the signer.
The *root* and the *cert* will be loaded only once either way.

I would be ready to live with that. 
I would not *load* the certificates when parsing the chain file.
I at all, I would do this in [def verify_certificate](https://github.com/privacyidea/privacyidea/blob/2152_attestation_parameter/privacyidea/lib/tokens/certificatetoken.py#L140) just loading the chain from pem strings to cert objects:

~~~~python
def verify_certificate(certificate, chain):
    certificate = load_pem_x509_certificate(to_byte_string(certificate), default_backend())
    chain = [load_pem_x509_certificate(to_byte_string(c), default_backend()) for c in chain]
    
~~~~

This way we could even omit the helper ``_verify_cert``.

Just drop a note, if you think it is necessary.

The new method would look like this:

~~~~python
def verify_certificate(certificate, chain):
    # first reverse the list, since it can be popped better
    chain = list(reversed(chain))
    if not chain:
        raise privacyIDEAError("Can not verify certificate against an empty chain.")
    certificate = load_pem_x509_certificate(to_byte_string(certificate), default_backend())
    chain = [load_pem_x509_certificate(to_byte_string(c), default_backend()) for c in chain]
    # verify chain
    while chain:
        signer = chain.pop()
        if chain:
            # There is another element in the list, so we check the intermediate:
            signee = chain.pop()
            signer.public_key().verify(
                signee.signature,
                signee.tbs_certificate_bytes,
                padding.PKCS1v15(),
                signee.signature_hash_algorithm
            )
            signer = signee

    # This was the last certificate in the chain, so we check the certificate
    signer.public_key().verify(
        certificate.signature,
        certificate.tbs_certificate_bytes,
        padding.PKCS1v15(),
        certificate.signature_hash_algorithm
    )
~~~~

Reply:
Done this in d6f3d8d5